File: mvcc_wal.go
package main

import (
	"fmt"
	"os"
)

// MVCC
type Version struct {
	value   int
	beginTx int
	endTx   int // 0 = infinity
}

// WAL
type WAL struct {
	file *os.File
}

func NewWAL(path string) *WAL {
	f, err := os.OpenFile(path, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
	if err != nil {
		panic(err)
	}
	return &WAL{file: f}
}

func (w *WAL) Log(record string) {
	if _, err := w.file.WriteString(record + "\n"); err != nil {
		panic(err)
	}
}

func (w *WAL) Sync() {
	if err := w.file.Sync(); err != nil {
		panic(err)
	}
}

// store
type Store struct {
	versions []*Version
	nextTx   int
	wal      *WAL
}

func NewStore() *Store {
	return &Store{
		versions: make([]*Version, 0, 1024),
		nextTx:   1,
		wal:      NewWAL("wal_go.log"),
	}
}

func (s *Store) Begin() int {
	tx := s.nextTx
	s.nextTx++

	// WAL: BEGIN
	s.wal.Log(fmt.Sprintf("BEGIN %d", tx))
	s.wal.Sync()

	return tx
}

func (s *Store) Read(tx int) int {
	for i := len(s.versions) - 1; i >= 0; i-- {
		v := s.versions[i]
		if v.beginTx <= tx && (v.endTx == 0 || v.endTx > tx) {
			return v.value
		}
	}
	return 0
}

func (s *Store) Write(tx int, value int) {
	// WAL: WRITE
	s.wal.Log(fmt.Sprintf("WRITE %d %d", tx, value))
	s.wal.Sync()

	// Apply MVCC update
	for i := len(s.versions) - 1; i >= 0; i-- {
		v := s.versions[i]
		if v.beginTx <= tx && (v.endTx == 0 || v.endTx > tx) {
			v.endTx = tx
			break
		}
	}

	s.versions = append(s.versions, &Version{
		value:   value,
		beginTx: tx,
	})
}

func main() {
	store := NewStore()
	const N = 1_000

	for i := 0; i < N; i++ {
		tx := store.Begin()
		store.Read(tx)
		store.Write(tx, i)
	}

	fmt.Println("Go MVCC & WAL done:", store.Read(store.nextTx-1))
}

File: MVCC_WAL.java
import java.io.*;
import java.nio.channels.FileChannel;
import java.util.*;

// MVCC
class Version {
    int value;
    int beginTx;
    int endTx; // 0 = infinity

    Version(int value, int beginTx) {
        this.value = value;
        this.beginTx = beginTx;
        this.endTx = 0;
    }
}

// WAL
class WAL {
    private final FileChannel channel;

    WAL(String path) {
        try {
            FileOutputStream fos = new FileOutputStream(path, true);
            this.channel = fos.getChannel();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    void log(String record) {
        try {
            byte[] bytes = (record + "\n").getBytes();
            channel.write(java.nio.ByteBuffer.wrap(bytes));
            channel.force(true);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    void sync() {
        try {
            channel.force(true);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

// Store
public class MVCC_WAL {
    List<Version> versions = new ArrayList<>();
    int nextTx = 1;
    WAL wal = new WAL("wal_java.log");

    int begin() {
        int tx = nextTx++;

        // WAL: BEGIN
        wal.log("BEGIN " + tx);
        wal.sync();

        return tx;
    }

    int read(int tx) {
        for (int i = versions.size() - 1; i >= 0; i--) {
            Version v = versions.get(i);
            if (v.beginTx <= tx && (v.endTx == 0 || v.endTx > tx)) {
                return v.value;
            }
        }
        return 0;
    }

    void write(int tx, int value) {
        // WAL: WRITE
        wal.log("WRITE " + tx + " " + value);
        wal.sync();

        // Apply MVCC update
        for (int i = versions.size() - 1; i >= 0; i--) {
            Version v = versions.get(i);
            if (v.beginTx <= tx && (v.endTx == 0 || v.endTx > tx)) {
                v.endTx = tx;
                break;
            }
        }

        versions.add(new Version(value, tx));
    }

    public static void main(String[] args) {
        MVCC_WAL store = new MVCC_WAL();
        final int N = 1_000;

        for (int i = 0; i < N; i++) {
            int tx = store.begin();
            store.read(tx);
            store.write(tx, i);
        }

        System.out.println("Java MVCC & WAL done: " + store.read(store.nextTx - 1));
    }
}
